hal/src/thumbv7em/clock/v2/xosc.rs

```diff
@@ -26,169 +26,125 @@ use typenum::U0;
+pub enum XoscId0 {}
 
-    impl SourceMarker for Xosc1 {}
+impl Sealed for XoscId0 {}
 
-    impl NotGclkInput for Xosc1 {}
+/// Type which serves as a source marker for the [`super::Xosc0`] and
+/// provides numerical identity for it
+impl XoscId for XoscId0 {
+    const NUM: usize = 0;
+    type XIn = PA14;
+    type XOut = PA15;
 }
 
-/// Encapsulates a hardware crystal configuration
+/// Type-level variant representing the identity of XOSC1
 ///
-/// Supported crystal frequencies: between 8 MHz and 48 MHz
+/// This type is a member of several [type-level enums]. See the documentation
+/// on [type-level enums] for more details on the pattern.
 ///
-/// User must provide the correct hardware frequency here,
-/// that allows configuring hardware to optimise for specific frequencies used.
-pub struct CrystalConfig {
-    src_freq: Hertz,
-    current: CrystalCurrent,
-}
+/// [type-level enums]: crate::typelevel#type-level-enum
+pub enum XoscId1 {}
 
-impl CrystalConfig {
-    /// Create a new [`CrystalConfig`] running at frequency `src_freq`
-    pub fn new(src_freq: impl Into<Hertz>) -> Result<Self, ()> {
-        // Calculate the CrystalCurrent from the supplied
-        // crystal frequency
-        let (current, src_freq) = match src_freq.into().0 {
-            freq @ 8_000_000 => (CrystalCurrent::BaseFreq8m, freq),
-            freq @ 8_000_001..=16_000_000 => (CrystalCurrent::LowFreq8mTo16m, freq),
-            freq @ 16_000_001..=24_000_000 => (CrystalCurrent::MedFreq16mTo24m, freq),
-            freq @ 24_000_001..=48_000_000 => (CrystalCurrent::HighFreq24mTo48m, freq),
-            _ => return Err(()),
-        };
-
-        Ok(Self {
-            src_freq: src_freq.hz(),
-            current,
-        })
-    }
-
-    /// Return the frequency of the crystal
-    pub fn freq(&self) -> Hertz {
-        self.src_freq
-    }
+impl Sealed for XoscId1 {}
 
-    /// Return the crystal current parameters based on frequency
-    pub fn get_current(&self) -> CrystalCurrent {
-        self.current
-    }
+impl XoscId for XoscId1 {
+    const NUM: usize = 1;
+    type XIn = PB22;
+    type XOut = PB23;
 }
 
-#[derive(Clone, Copy, PartialEq)]
-/// Current mutliplier/reference pair
+//==============================================================================
+// CrystalCurrent
+//==============================================================================
+
+/// Crystal current settings
+///
+/// The crystal current fields (`IMULT` and `IPTAT`) are usually set by the
+/// crystal frequency range. Normally:
+/// - 8 MHz -> `Low`      
+/// - 8 MHz to 16 MHz -> `Medium`   
+/// - 16 MHz to 24 MHz -> `High`     
+/// - 24 MHz to 48 MHz -> `ExtraHigh`
+///
+/// However, the datasheet notes that the next lower setting can be used if the
+/// capacitive load is particularly small, which would save power.
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[allow(missing_docs)]
 pub enum CrystalCurrent {
-    /// 8MHz
-    BaseFreq8m,
```

Comment:
I'm not married to the names. But I wanted to avoid something like this:
```rust
let config = CrystalConfig::new(25.mhz());
let xosc = Xosc::from_crystal(token, xin, xout, config);
    .set_current(CrystalCurrent::LowFreq8mTo16m)
    .enable();
```
This code is perfectly reasonable and valid for a 25 MHz oscillator with a low capacitive load. Yet the code seems to directly contradict itself. In one line, it says the frequency is 25 MHz. Then in another, it says the frequency is in the range 8 MHz to 16 MHz.

If some other user came along and read this code later, I think it would be seen as a possible bug. At the very least, it's confusing if you don't already know the datasheet well.

I don't particularly care what the variant names are. But I would prefer them not to contain frequencies, because the frequencies are only *suggestions* by the datasheet.

Reply:
Ok, makes sense